lib/Sema/TypeCheckMacros.cpp

```diff
@@ -338,11 +343,30 @@ loadExecutablePluginByName(ASTContext &ctx, Identifier moduleName) {
     if (fs->getRealPath(libraryPath, resolvedLibraryPath)) {
       return nullptr;
     }
+    std::string resolvedLibraryPathStr(resolvedLibraryPath);
+    std::string moduleNameStr(moduleName.str());
+
     bool loaded = swift_ASTGen_pluginServerLoadLibraryPlugin(
-        executablePlugin, resolvedLibraryPath.c_str(), moduleName.str().data(),
+        executablePlugin, resolvedLibraryPathStr.data(), moduleNameStr.data(),
         &ctx.Diags);
     if (!loaded)
       return nullptr;
+
+    // Set a callback to load the library again on reconnections.
+    auto *callback = new std::function<void(void)>(
+        [executablePlugin, resolvedLibraryPathStr, moduleNameStr]() {
+          (void)swift_ASTGen_pluginServerLoadLibraryPlugin(
+              executablePlugin, resolvedLibraryPathStr.data(),
+              moduleNameStr.data(),
+              /*diags=*/nullptr);
+        });
+    executablePlugin->addOnReconnect(callback);
+
+    // Remove the callback and deallocate it when this ASTContext is destructed.
+    ctx.addCleanup([executablePlugin, callback]() {
+      executablePlugin->removeOnReconnect(callback);
```

Comment:
This is _not_ safe when the `PluginRegistry` is owned by `ASTContext`. Currently when `ASTContext` is to own a `PluginRegistry`, it adds a cleanup function like `this->addCleanup([registry]{ delete registry; })`, and cleanup functions are called in the order they are added. That means that the plugin registry and its plugins would be deleted before this callback is called. To resolve this:
* Instead of using cleanup callback, Add a field`std::unique_ptr<PluginRegistry> OwnedPluginRegistry` to `ASTContext::Implementation` so it'd be alive until these callbacks are called
* or, call cleanup callbacks in reverse order, like a stack

@bnbarham wdyt?

Reply:
Went with the former because I feel we should reduce `addCleanup` usages whenever possible.

Reply:
Sounds good to me. I'm all for more managed pointers :). IMO the latter also makes sense as a general change.